crates/dip/src/lib.rs

```diff
@@ -45,13 +45,27 @@ impl<Identifier, Proof, Details> From<v1::IdentityProofAction<Identifier, Proof,
 }
 
 #[derive(Encode, Decode, RuntimeDebug, Clone, Eq, PartialEq, TypeInfo)]
-pub enum VersionedIdentityProof<LeafKey, LeafValue> {
+pub enum VersionedIdentityProof<BlindedValue, Leaf> {
```

Comment:
The idea is that consumers would try to convert their own version into this for message passing, and receivers would try to convert it back into a specific version. I took inspiration for how [VersionedXcm](https://github.com/paritytech/polkadot/blob/master/xcm/src/lib.rs#L376) works. Why do you think we need `non_exhaustive`?

Reply:
I was thinking how a new version would be introduced. What happens if a VersionedIdentityProof::v2 get's send to someone that only supports v1? at the moment that would result in a decoding error. Adding `non_exhaustive` wouldn't change anything here.

But if you use the `dip` crate as a dependency and use the `VersionedIdentityProof` having `non_exhaustive` makes a difference. If `VersionedIdentityProof` is not marked as `non_exhaustive`, adding a new version is a breaking change (since `match VersionedIdentityProof {}` doesn't cover this new case). Having `non_exhaustive` makes adding new versions a non-breaking change since everyone is forced to already have code in place that covers new versions (and showing an error that this version is not supported).

Reply:
I see. But isnt'a a compilation error what we actually want here? If you risk getting a versioned identity proof you don't support, you should probably return an error. Code won't need to be updated if you are using the old version, hence anything failing to decode will result in an error. Once you update the dependency, you should be at least aware that a new version could be around, and explicitly generate an error if it's not what you expect. Or am I still missing your point?

Reply:
The `non_exhaustive` is only for compile time. If we introduce a new version this would be a breaking change. Since it fails to compile.

At runtime this doesn't change anything since if we talk to a newer version that we don't know, it fails to decode and won't reach any match statement where we check versions.

Reply:
https://github.com/KILTprotocol/kilt-node/pull/492/commits/343faf908b60371a9b48ae4a75047016c366169e. Let's see how this plays out 😄